opal/mca/accelerator/rocm/accelerator_rocm_component.c

```diff
@@ -196,41 +197,94 @@ int opal_accelerator_rocm_lazy_init()
+    }
     *opal_accelerator_rocm_MemcpyStream = memcpy_stream;
 
     opal_accelerator_rocm_mem_bw = malloc(sizeof(float)*opal_accelerator_rocm_num_devices);
+    if (NULL == opal_accelerator_rocm_mem_bw) {
+        opal_output(0, "Could not allocate memory_bw array\n");
+        err = OPAL_ERR_OUT_OF_RESOURCE;
+        goto out;
+    }
+
     for (int i = 0; i < opal_accelerator_rocm_num_devices; ++i) {
         int mem_clock_rate; // kHz
-        err = hipDeviceGetAttribute(&mem_clock_rate,
-                                    hipDeviceAttributeMemoryClockRate,
-                                    i);
+        hip_err = hipDeviceGetAttribute(&mem_clock_rate,
+                                        hipDeviceAttributeMemoryClockRate,
+                                        i);
+        if (hipSuccess != hip_err) {
+            opal_output(0, "Failed to query device MemoryClockRate, err=%d %s\n",
+                        hip_err, hipGetErrorString(hip_err));
+            err = OPAL_ERROR;
+            goto out;
+        }
+
         int bus_width; // bit
-        err = hipDeviceGetAttribute(&bus_width,
-                                    hipDeviceAttributeMemoryBusWidth,
-                                    i);
-        /* bw = clock_rate * bus width * 2bit multiplier
-         * See https://forums.developer.nvidia.com/t/memory-clock-rate/107940
-         */
+        hip_err = hipDeviceGetAttribute(&bus_width,
+                                        hipDeviceAttributeMemoryBusWidth,
+                                        i);
+        if (hipSuccess != hip_err) {
+            opal_output(0, "Failed to query device MemoryBusWidth, err=%d %s\n",
+                        hip_err, hipGetErrorString(hip_err));
+            err = OPAL_ERROR;
+            goto out;
+        }
+
+        /* bw = clock_rate * bus width * 2bit multiplier */
         float bw = ((float)mem_clock_rate*(float)bus_width*2.0) / 1024 / 1024 / 8;
         opal_accelerator_rocm_mem_bw[i] = bw;
     }
+
+#if HIP_VERSION >= 60000000
+    int dev_id;
+    hip_err = hipGetDevice(&dev_id);
+    if (hipSuccess != hip_err) {
+        opal_output(0, "error retrieving current device");
+        err = OPAL_ERROR;
+        goto out;
+    }
+
+    int has_large_bar = 0;
+    hip_err = hipDeviceGetAttribute (&has_large_bar, hipDeviceAttributeIsLargeBar,
+                                     dev_id);
+    if (hipSuccess != hip_err) {
+        opal_output(0, "error retrieving current device");
+        err = OPAL_ERROR;
+        goto out;
+    }
+
+    if (0 < has_large_bar) {
+        // Without large BAR we have to use hipMemcpy(Async) for all data transfers
+        opal_output(0, "Large BAR support is not enabled on current device. "
+                    "Enable large BAR support in BIOS (Above 4G Encoding) for "
+                    "better performance\n.");
+        opal_accelerator_rocm_memcpyH2D_limit = 0;
+        opal_accelerator_rocm_memcpyD2H_limit = 0;
+    }
```

Comment:
I don't understand what a value of `has_large_bar` larger than 0 means? That there it does not have large BAR? Couldn't find any documentation about it.

Also, should we limit the output to higher verbosity levels to not spam users?

Reply:
Argh, thank you for catching it, that should be == 0 ( I set to > zero to trigger the output once, and forgot to reset it)

Reply:
regarding the level, I think this is quite important, I would prefer that the user has that on their screen. Until this PR, we segfaulted if it was not set :-(